### DISCUSSION TITLE
Variables: Can't use space in variable search

### DISCUSSION THREAD

**Discussion by tobiasdirksen** [grafana/grafana#38173] [Dashboards, Panels & Visualizations]
**What happened**:
Since one of the recent updates (I'm not sure which one), I can't use space anymore, when I search for a value option in a variable. This is a big problem, e.g., when you have many addresses in the variable, and you would like to search for a specific one. When I click space nothing happens. 

**What you expected to happen**:
I expect that a space is inserted in the search string, when I press space.

**How to reproduce it (as minimally and precisely as possible)**:
Create a custom variable with options including space on an empty dashboard:

![billede](https://user-images.githubusercontent.com/7211545/98226586-88574000-1f56-11eb-9513-844188cb9b5e.png)
======
Then try to search for one of the options. I can't insert space:
![billede](https://user-images.githubusercontent.com/7211545/98226734-b76db180-1f56-11eb-9b6e-3f2642d7606e.png)

**Anything else we need to know?**:
It does actually work on Android 5.1.1 and Chrome 86.0.4240.110:
![billede](https://user-images.githubusercontent.com/7211545/98228434-e127d800-1f58-11eb-9eea-55ef0cddc32c.png)

**Environment**:
- Grafana version: 7.3.1
- OS Grafana is installed on: Ubuntu 18.04.5 LTS
- User OS & Browser: Windows 10 Enterprise 18.03. Edge (86.0.622.61)/Firefox (82.0.2)/Chrome (86.0.4240.111)
- Grafana plugins: No plugin is used

**Comment by marefr** [comment]
Thanks. Please update your issue description - use the bug template to provide more information and needed details https://github.com/grafana/grafana/issues/new?labels=type%3A+bug&template=1-bug_report.md

**Comment by torkelo** [comment]
Unable to replicate 
<img width="250" alt="Screen Shot 2020-11-06 at 07 50 24" src="https://user-images.githubusercontent.com/10999/98335164-c317b180-2004-11eb-92e9-44bdc9ffbeb3.png">

**Comment by tobiasdirksen** [comment]
Thanks for quick reply. Hm, all my colleagues have the same problem. It's probably related to user OS or browser version then.

**Comment by torkelo** [comment]
But there is a problem inserting space, sorry I miss read this

**Comment by torkelo** [comment]
This is by design space is used to select / deselect values

**Reply by shanielh** [reply]
Also for single value variables?

**Comment by tobiasdirksen** [comment]
Thanks. I see. However, this has worked previously, and does work on Android, and also on your screenshot above? I don't think it is that common to select with space in a dropdown, but maybe I'm wrong. Is there any workaround besides pasting the whole search text into the search field?

**Comment by DimaRabkin** [comment]
+1

I upgraded version 7.0.3 to 7.3.1 and it stopped working.
I checked this with Safari and Chrome

**Comment by tobiasdirksen** [comment]
I could maybe also add that space doesn't select/deselect in a single value variable. Nothing happens when I press space. It only works for multivalued variables, where it is also more intuitive, in my opinion.

**Comment by Grrrolf** [comment]
> This is by design space is used to select / deselect values

A suggestion: 
Instead of pressing space in the template variable search box for making a selection, what about having selecting and deselecting using space only work when highlighting the actual value (using the mouse or cursor keys). In other wordt, scroll down to the value which needs to be selected or deselected and press space. This will allow for quickly making one or even multiple selection.

In our dashboards showing for customers or products metrics, we fill the template variables with a list of customers or products and let the users use the template variable search box to search for a customer or products. This works really great and quick even, when the list of values is rather large.

**Comment by mpaloni** [comment]
+1 Copy pasting works as said, but would be nice to write the space directly to search field.

**Comment by jdburche** [comment]
While testing this functionality in Firefox and Chrome (running on Mac OS 11) on Grafana v7.3.7 I observed the following:

  * The spacebar **does** select items from multi-value drop down lists for variables
  * The spacebar **does not** select items from drop down lists for variables that are not multi-value
 
It would be nice to have spaces considered when searching these lists. If nothing else, perhaps this behavior could be made configurable for each variable?

**Comment by uebenes** [comment]
I have the same problem. You can not type any spaces to filter the result set. That is really confusing.

**Comment by barts2108** [comment]
We use Grafana where we present the user with a dropdown of names that describe digital and analog IO's (the real IO names are something like AI0000, AI0001 etc

![no whitespace](https://user-images.githubusercontent.com/520875/130081211-05732481-c973-4d4d-a34a-31942ed028f0.PNG)

In version 7.1.1 the whitespace did select an item in the dropdown and put the space into the input box as well. Now in version 8.1.1 the whitespace key is compleletly blocked. We would love to see the possibility to simply type in the searchbox (including whitespace) where even in the multiselect there would be an option that the spacebar does **not** select any item. Only use it for filtering.

Does anyone know where in the code the whitespace character is blocked ?

**Comment by pkolyvas** [comment]
We're in the process of moving Grafana feature requests to [GitHub Discussions](https://community.grafana.com/t/were-moving-grafana-feature-requests-to-github-discussions/50341/3). This issue is being moved to [Discussions](https://github.com/grafana/grafana/discussions) or has been moved from our issues list, depending on where you're seeing this message. 

 We hope to use the discussion thread to continue building use cases, and additional evidence based on engagement/voting, in order to help us prioritize this feature request later on down the road. 

As with all things Grafana, we value your input and participation in this project and appreciate your patience during this transitional period. Looking forward to the discussion!

**Reply by barts2108** [reply]
Understood. Elaborating on our use-case: Our metric names are dumb single character IO abbreviations (A for analog, D for digital) with a second character for input (I) or output (O). These characters are followed with a number ranging from 0 to 9999. In our variable query we use a lookup table that links the IO name to a user understandable name (as shown in the picture in my previous post).

With Grafana 7.1.1 the space bar was selecting the highlighted item **and** the space as added to the variable search input box. There we could use the sequence "spacebar - backspace - spacebar" to add the space to the search input box while the highlighted item was selected and unselected. 

Now with Grafana 8.1.1 the space bar is completely blocked which is a problem for our search.

It would be incredible usefull if it would be configurable, where we could configure to 
(1) spacebar selects highlighted item - true/false 
(2) spacebar is blocked in searchbox - true/false

Currently I am searching for the location where the space bar is blocked for the search input box, but so far I did not find that location (yet). If I know where this space bar is handled, I might be able to find some solution.

**Comment by barts2108** [comment]
This story seems to have even more quirks. 

When a template variable is clicked (input box has focus) and the dropdown is shown, it can be navigated with up/down arrow keys. However if the list in the dropdown is longer than shown (there is a scrollbar), moving down more with the arrow down key does **not** scroll the list, the highlighted item dissappears out of the visible part. When the dropdown is clicked and the arrow/up down is used, the list is scrolled but the highlighted items stays the same.

When the cursor is in the input box, and you press pageup or pagedown key, the entire page is scrolled, where I expected the drop down to scroll a page up or down. Home and End key move the cursor to the front or the end of the text in the input box, and not the dropdown to the top or all to the bottom.

When an item in the dropdown is clicked (either selected or unselected), home, end, pgup and pgdn keys work as expected in the dropdown, but the space bar now works as a page down as well.

When pressing ENTER in the inputbox where characters exsit (and a dropdown item is highlighted but not selected), the highlighted item is selected **and** the drop down is closed. When you open the dropdown again, and press enter in an empty input box, all selected items are unselected. **BUT** when you have a selected items and enter a search in the input box that highlights the selected item, pressing ENTER will close the dropdown but not unselect the selected item.

What I would have expected: 

- Arrow up/down to scroll the drop down when the highlighted items moves out of sight

- PgUp / PgDn to scroll the drop down when it is opened **and** move the highlighted item

- Home / End to scroll the drop down when it is opened **and** move the hightlighted item

- Enter key to be consistent in selecting/unselecting 

For the space key
- When the input box has focus, the space should be use to filter (if some metrics may not have whitespaces, this might be something that should be configurable)
- When the dropdown has focus, the space bar should select/unselect the highlighted (hovered) item, not scroll the list.

For further info I added our variable query (we use MSSQL) and the panel query. 

We have a table containing metrics (DATA_LOG) and a lookup table (DATA_LOG_LOOKUP) which are linked throught the field `point_id`. This allows us to have "human readable names with whitespaces" for the dropdown, and "point_id's" for the metrics query.  Furthermore we did some tweaking to secure the query against a division by 0, and we get more points than are shown on the screen (which does some kind of fill previous / fill next in the panel)

All of it works very well. Except the variable drop down not accpeting whitespaces to search is the main (and breaking) issue.

Dashboard variable `anyIO`

`SELECT  point_name as __text, point_id as __value from DATA_LOG_LOOKUP WHERE point_id LIKE 'A%' OR  point_id LIKE 'D%'`

```
DECLARE @maxpoints FLOAT
DECLARE @timeFrom   DATETIME2
DECLARE @timeTo     DATETIME2
DECLARE @timeDiff   INTEGER
DECLARE @timeOffset FLOAT

SET @timeOffset = 0.01

SET @maxpoints = $__interval_ms / 10000

IF @maxpoints < 1.0
BEGIN
  SET @maxpoints = 1.0
END


SET @timeDiff = DATEDIFF(ss, $__timeFrom(), $__timeTo())
IF (@timeDiff < (2*60))
BEGIN
  SET @timeDiff = (2*60)
END

SET @timeFrom = DATEADD(ss, -(@timeDiff * @timeOffset), $__timeFrom())
SET @timeTo   = DATEADD(ss, (@timeDiff * @timeOffset), $__timeTo())


SELECT   FLOOR(DATEDIFF_BIG(millisecond,'1970-01-01 00:00:00.000', timestamp_utc)/@maxpoints)*@maxpoints as time,
         l.point_name as metric,
         avg(CAST(_VAL as FLOAT)) as value
FROM     DATA_LOG d
JOIN     DATA_LOG_LOOKUP l
ON       l.point_id = d.point_id
WHERE (
         d.point_id IN(${anyIO}​) 
         AND
         timestamp_utc BETWEEN @timeFrom AND @timeTo )
GROUP BY FLOOR(DATEDIFF_BIG(millisecond,'1970-01-01 00:00:00.000', timestamp_utc)/@maxpoints)*@maxpoints, l.point_name
ORDER BY time, metric
```

Tested with Grafana 8.1.2 (and 7.1.1)
OS : Windows 10
Browsers: Edge (92.0.902.78 (Official build) (64-bit)), FireFox 91.0.1 (64-bit), CefSharp.WinForms 91.1.230

**Comment by daldoyle** [comment]
It would be great if this could be addressed somehow, as I and others I've spoken to feel like the current behavior is unintuitive in all cases. Not being able to enter the space character while a textbox has focus is confusing for end users and breaks the ability to search for terms where the whitespace is important. Pressing the spacebar to select items in a dropdown is a standard browser behavior but only when the item in question is being hovered over. Anecdotally speaking, I have never thought to use the spacebar to select an item from a dropdown in Grafana fwiw. There have been a number of good ideas already posted here.

**Comment by barts2108** [comment]
Because we can't wait for a decent solution, I have taken the freedom to check if I could implement some hacky way to achieve this. And I succeeded. 

At the end of the "grafana\public\build\vendors~ app.8ea303538f79ee32b68a.js" file (I used Grafana v8.1.2 (103f8fa094) on windows), but before this line ";//# sourceMappingURL=vendors~app.8ea303538f79ee32b68a.js.map", add the following code.

- this must be done after each update/install/reinstall of grafana
- this does not work is a space is used as very first character
- works with multiple variables

If this helps someone to save millions, feel free to donate some of it to me :laughing: :grinning: 

```
if (typeof jQuery === 'undefined') {
  // jQuery is NOT available
  alert("no jquery");
} else {
$('body').on('keydown', function(e){
  if (e.target.className == 'gf-form-input') {
	installHack(e);
  }
});
}
installHack = function(e) {
  $(e.target).prop('onkeydown', 'hackAllowSpace(event);');
  addHack(e);
}
hackAllowSpace = function(e) {
  e = e || window.event;
  // user has pressed space
  if(e.which == 32){
	  // Get cursor position and string length
	  var position = e.target.selectionStart;
	  var strlength = e.target.value.length;
	  
	  // prevent default will prevent keypres from happening and causes
	  // that the list is not filtered. therefore commented here!
	  // side effect is that keypress event still adds a space at the
	  // end of the input...
          //e.preventDefault();
	  e.stopPropagation();
	  
	  // Add the space at the correct position (in between the characters or at the end
	  var newVal = $(e.target).val().substring(0, position) + ' ' + $(e.target).val().substring(position);
	  
	  //console.log('targetval1:' + $(e.target).val() + ':');
	  //console.log('newval    :' + newVal + ':');
	  
	  // Set the new value to the search box
          $(e.target).val(newVal);
	  
	  //console.log('targetval2:' + $(e.target).val() + ':');
	  
	  //NASTY! Use a setTimeout to (first) let keypress add the 
	  // whistspace at the end of the input, then upon timer 
	  // expiration remove that whitespace again.
	  setTimeout(function() {
		var strlength = e.target.value.length;
		var newVal = $(e.target).val().substring(0, strlength - 1)
		$(e.target).val(newVal);
		//console.log('targetval3:' + $(e.target).val() + ':');
		e.target.setSelectionRange(position+1,position+1);
	  },1, e, position);
  }
}
addHack = function(e) {
  var elem = e.target;
  var oldevent = elem.onkeydown;
  if (oldevent) {
    elem.onkeydown = function() {
      hackAllowSpace();
      oldevent();
    }
  } else {
    elem.onkeydown = function() {
      hackAllowSpace();
    }
  }
}
```

**Reply by uebenes** [reply]
@barts2108:
 
In Grafana 8.3.4 the file "grafana\public\build\vendors~ app.*.js" is not existing anymore. I think I found the right file (5105.1322a38d8e955d3d27e7.js) and insert your code. But it seems, that this hack is not working anymore. Is there any solution for the current Grafana version?

**Reply by barts2108** [reply]
Well I did have the same problem in 8.3.3 and I am not sure if 8.3.4 will work.

I placed my hack in `DashboardPage.1f65f885.....js`
Just before the `//# sourcemapping` line

**Comment by rodrigo-brito** [comment]
Same problem here, the problem still occurring in version 8.2

**Comment by barts2108** [comment]
I did have the same problem as @uebenes and solved it in 8.3.3. I am not sure if 8.3.4 will work.

I placed my hack in DashboardPage.1f65f885.....js
Just before the //# sourcemapping line

**Comment by murtazaqa** [comment]
I am getting this issue in v9.3.0 as well. Unable to filter by space, it actually selects the first value in the list.

**Comment by bgeels** [comment]
Also encountering this issue. Are there any plans to resolve this? It seems that this is pretty clearly a bug. The intention of the spacebar shortcut for drop down item selection is to only work after a drop down item has been navigated to by way of arrow keys or some other shortcut, not to interrupt the user's search query by selecting the first item as they're still typing the search term.

**Comment by natellium** [comment]
Hi everyone, this was on our radar (https://github.com/grafana/grafana/issues/60589) but had missed the discussion. We did considered it indeed a bug. 

It was fixed in [this](https://github.com/grafana/grafana/pull/60622) was released as part of 9.4. Nevertheless, we later discovered the fix introduced a regression that we're fixing [here](https://github.com/grafana/grafana/pull/63495) and we're aiming to release with the next patch. 

I will lock the discussion. If you have any more feedback regarding this please do open a bug in our repo :)
